std/range/primitives.d

```diff
@@ -1688,7 +1688,7 @@ The following expression must be true for `hasSlicing` to be `true`:
  */
 enum bool hasSlicing(R) = isForwardRange!R
     && !(isAutodecodableString!R && !isAggregateType!R)
-    && is(typeof((R r) { return r[1 .. 1].length; } (R.init)) == size_t)
+    && is(typeof(() { return lvalueOf!R[1 .. 1].length; } ()) == size_t)
```

Comment:
This adds a template, why do this here? Also, the inout problem is here too.

Reply:
OK, reverted. I thought using lvalueOf!R one more time in hasSlicing wouldn't matter as it is already there.

BTW on the forum you wrote:
> since isInputRange validates that typeof(R.init) == R, I just used R.init as the parameter

hasSlicing does not validate that, is it still OK to use R.init?

Reply:
`isForwardRange` calls `isInputRange`, which does the validation, so yes.

Reply:
>  I thought using lvalueOf!R one more time in hasSlicing wouldn't matter as it is already there.

Yes, this is a good point. I always forget that. `lvalueOf` is pretty much free anywhere in here, because it's used all over the place. It's one of the counter-intuitive properties of templates.